api/apis/service_instance_handler.go

```diff
@@ -183,8 +186,48 @@ func (h *ServiceInstanceHandler) serviceInstanceListHandler(authInfo authorizati
 	writeResponse(w, http.StatusOK, presenter.ForServiceInstanceList(serviceInstanceList, h.serverURL, *r.URL))
 }
 
+func (h *ServiceInstanceHandler) serviceInstanceDeleteHandler(authInfo authorization.Info, w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	vars := mux.Vars(r)
+	serviceInstanceGUID := vars["guid"]
+
+	serviceInstance, err := h.serviceInstanceRepo.GetServiceInstance(ctx, authInfo, serviceInstanceGUID)
+	if err != nil {
+		if repositories.IsForbiddenError(err) {
+			h.logger.Error(err, "user not allowed to get service instance")
+			writeNotFoundErrorResponse(w, repositories.ServiceInstanceResourceType)
+
+			return
+		}
+
+		h.logger.Error(err, "failed to get service instance")
+		writeUnknownErrorResponse(w)
+		return
+	}
+
+	err = h.serviceInstanceRepo.DeleteServiceInstance(ctx, authInfo, repositories.DeleteServiceInstanceMessage{
```

Comment:
Generally, the `Get-Delete` pattern when dealing with K8S is racy. I wonder whether we can implement the handler just call the repository's `Delete` method and handle an eventual not found error appropriately. The only thing the service instance found is used for is to get the space GUID, but on the other hand, the service instance GUID should be unique and therefore sufficient to locate and delete it.

Reply:
That is a very good point @tcdowney I will address it today. This is also a very good example of why we need to handle our errors in a more generic way to avoid overlooking corner cases like this one in future.